vdr/store/bbolt.go

```diff
@@ -0,0 +1,305 @@
+func (store *bboltStore) Iterate(fn vdr.DocIterator) error {
+	return store.db.View(func(tx *bbolt.Tx) error {
+		versions := tx.Bucket(versionsBucket)
+		if versions == nil {
+			return nil
+		}
+
+		documents := tx.Bucket(documentsBucket)
+		if documents == nil {
+			return nil
+		}
+
+		if err := versions.ForEach(func(key, data []byte) error {
+			versionList := &documentVersionList{}
+
+			if err := json.Unmarshal(data, versionList); err != nil {
+				return err
+			}
+
+			id, _ := did.ParseDID(string(key))
+			doc, err := store.getDocumentVersion(documents, *id, versionList.Latest())
+			if err != nil {
+				return err
+			}
+
+			if err := fn(doc.Document, doc.Metadata); err != nil {
+				return err
+			}
+
+			return nil
+		}); err != nil {
+			return err
+		}
+
+		return nil
+	})
+}
+
+func (store *bboltStore) filterDocument(doc *documentVersion, metadata *vdr.ResolveMetadata) error {
+	// Verify deactivated
+	if IsDeactivated(doc.Document) {
+		return vdr.ErrDeactivated
+	}
+
+	if metadata == nil {
+		return nil
+	}
+
+	// Verify creation and update time
+	if metadata.ResolveTime != nil {
+		resolveTime := *metadata.ResolveTime
+
+		if doc.Metadata.Created.After(resolveTime) {
+			return vdr.ErrNotFound
+		}
+
+		if doc.Metadata.Updated != nil {
+			if doc.Metadata.Updated.After(resolveTime) {
+				return vdr.ErrNotFound
+			}
+		}
+	}
+
+	// Verify KeyTransaction
+	if metadata.SourceTransaction != nil {
+		for i, keyTx := range doc.Metadata.SourceTransactions {
+			if keyTx.Equals(*metadata.SourceTransaction) {
+				break
+			}
+
+			if i == len(doc.Metadata.SourceTransactions)-1 {
+				return vdr.ErrNotFound
+			}
+		}
+	}
+
+	return nil
+}
+
+// Resolve returns the DID Document for the provided DID
+func (store *bboltStore) Resolve(id did.DID, metadata *vdr.ResolveMetadata) (document *did.Document, documentMeta *vdr.DocumentMetadata, txErr error) {
+	txErr = store.db.View(func(tx *bbolt.Tx) error {
+		documents := tx.Bucket(documentsBucket)
+		if documents == nil {
+			return vdr.ErrNotFound
+		}
+
+		// If `metadata.Hash` is set we can fetch the document without iterating
+		if metadata != nil && metadata.Hash != nil {
+			doc, err := store.getDocumentVersion(documents, id, *metadata.Hash)
+			if err != nil {
+				return nil
+			}
+
+			if err := store.filterDocument(doc, metadata); err != nil {
+				return err
+			}
+
+			document = &doc.Document
+			documentMeta = &doc.Metadata
+
+			return nil
+		}
+
+		versions := tx.Bucket(versionsBucket)
+		if versions == nil {
+			return vdr.ErrNotFound
+		}
+
+		versionList := &documentVersionList{}
+
+		data := versions.Get([]byte(id.String()))
+		if data == nil {
+			return vdr.ErrNotFound
+		}
+
+		if err := json.Unmarshal(data, versionList); err != nil {
+			return err
+		}
+
+		for _, versionHash := range versionList.Versions {
+			data = documents.Get(store.getDocumentKey(id, versionHash))
+			if data == nil {
+				continue
```

Comment:
this would be very odd?

Reply:
Indeed. But even then I'd argue that we shouldn't rely on `data` to be non-`nil` here.